Add a vector_calculus class

semi-h opened the pull request:
This PR moves the vector calculus operations currently implemented inside solver class out to a dedicated new class. This new structure should allow accessing vector calculus operations outside the solver class. Examples include the laplacian operation iterative Poisson solver needs and curl/divergence for postprocessing/monitoring.

semi-h commented on src/backend.f90:
```diff
@@ -67,12 +67,12 @@ end subroutine transeq_ders
-         !! derivatives are decided at runtime. Backend implementations
-         !! are responsible from directing calls to transeq_ders into
-         !! the correct algorithm.
+    subroutine tds_solve(self, du, u, tdsops)
```
Removed dirps argument here, it had become redundant when we introduced mesh class a while ago.

Nanoseb commented on src/vector_calculus.f90:
```diff
@@ -0,0 +1,346 @@
+  subroutine curl(self, o_i_hat, o_j_hat, o_k_hat, u, v, w, &
+                  x_der1st, y_der1st, z_der1st)
+    !! Curl of a vector field (u, v, w).
+    !! Inputs from velocity grid and outputs to velocity grid.
```
Shall we use the `CELL`/`FACE` etc. nomenclature instead of velocity/pressure grids? Just to be consistent and general

Nanoseb commented on src/vector_calculus.f90:
```diff
@@ -0,0 +1,346 @@
+                            y_stagder_v2p, y_interpl_v2p, &
+                            z_stagder_v2p, z_interpl_v2p)
+    !! Divergence of a vector field (u, v, w).
+    !! Inputs from velocity grid and outputs to pressure grid.
```
Same, convert to `CELL` and `FACE` data location instead of velocity/pressure

Nanoseb commented on src/vector_calculus.f90:
```diff
@@ -0,0 +1,346 @@
+                          y_stagder_p2v, y_interpl_p2v, &
+                          z_stagder_p2v, z_interpl_p2v)
+    !! Gradient of a scalar field 'pressure'.
+    !! Inputs from pressure grid and outputs to velocity grid.
```
same as above

Nanoseb commented on src/vector_calculus.f90:
```diff
@@ -0,0 +1,346 @@
+    u_y => self%backend%allocator%get_block(DIR_Y)
+    d2u_y => self%backend%allocator%get_block(DIR_Y)
+
+    call self%backend%reorder(u_y, u, RDR_X2Y)
```
is it worth checking that `u` is in `DIR_X` before doing that? or at least add a comment in the header to explain that it is expected

semi-h replied:
`tds_solve` is doing a check based on directions so added a comment in the laplacian subroutine header to explain

Nanoseb reviewed (commented):
looks good to me, I just think it would be good to properly define what is expected from the input variables in terms of direction and dataloc.

semi-h commented:
> looks good to me, I just think it would be good to properly define what is expected from the input variables in terms of direction and dataloc.

Thanks! Changed the namings from velocity/pressure grid to data_loc based VERT/CELL. Also added a bit more explanation in subroutine headers to explain expected `DIR`s and `data_loc`s for input and output arrays.

pbartholomew08 commented on src/backend.f90:
```diff
@@ -67,12 +67,12 @@ end subroutine transeq_ders
+      !! the correct algorithm.
       import :: base_backend_t
       import :: field_t
       import :: dirps_t
```
Based on the above comment, I think we can get rid of this `import :: dirps_t` too?

pbartholomew08 commented on src/vector_calculus.f90:
```diff
@@ -0,0 +1,368 @@
+
+  type :: vector_calculus_t
+    !! Defines vector calculus operators
+    class(base_backend_t), pointer :: backend
```
This is the kind of circular pointer that I think we should try and avoid (i.e. a pointer to the `backend` object in the parent class of `vector_calculus`). Could we pass `backend` as an argument to `vector_calculus` methods? e.g.
```
call vector_calculus%curl(..., backend)
```

semi-h replied:
We don't really have a circular pointer or dependency here. `backend`, `solver` and `vector_calculus` are all independent classes and there is no inheritance relationship among them. There are other classes too that use `backend`, like `time_integrator` and new wind turbine stuff. In a class diagram `backend` would be under all of them so I don't see a problem having a pointer to `backend` in these classes. Why do you think we should avoid?

pbartholomew08 commented:
> Nanoseb: is it worth checking that u is in DIR_X before doing that? or at least add a comment in the header to explain that it is expected
> semi-h: tds_solve is doing a check based on directions so added a comment in the laplacian subroutine header to explain

I think it might be worth checking the conditions at the start of the curl/div/grad/xxx subroutines. A "invalid field locations passed to curl" will be easier to debug than "invalid fields in tdsops" (which you then need to figure out where it came from). Having the tdsops perform this last level check is good to ensure we catch everything, but a higher-level check might ultimately be beneficial in my opinion.
